consensus/enclave/impl/src/constant_time_token_map.rs

```diff
@@ -0,0 +1,331 @@
+//! This module defines a helper object for tracking and accumulating fees
+//! in the enclave, while remaining constant time with respect to token ids.
+//! This is used both when looking up the fee for a transaction, and when
+//! accumulating all the fees when forming a block.
+//!
+//! Note, in the future it may be more efficient to use data-structures / code
+//! from the mc-oblivious repo, which is optimized for x86-64.
```

Comment:
I'm guessing here you're basically talking about the difference between using a simple-ish CT approach vs. Oram. 

In napkin calculation terms, how much speedup/resource savings would you'd imagine we'd see by using oram vs. constant time? I'm not %100 sure how to quantify the measurement domain (i.e. per block or request to FOG maybe?) to quantify these speedup/savings either, so perhaps you can comment on that too?

Also what kind of implementation difficulty would we see using that paradigm over the constant time one?

Reply:
> I'm guessing here you're basically talking about the difference between using a simple-ish CT approach vs. Oram.

I guess there's three possible steps

1. Simple-ish CT (like in this PR)
2. Use aligned-cmov instead of subtle (which is going to use inline assembly, instead of the stuff in the `ConditionallySelectable` trait)
3. Use ORAM

So the big picture here is, we use this structure when:
* Checking if a transaction paid the minimum fee
* Accumulating fees during form_block

We check if a transaction paid the minimum fee when checking if a `Tx` is `WellFormed`. That already entails a lot of elliptic curve operations, checking bullet proofs, etc. etc. So that's probably on the order of like 100s of microseconds or like a millisecond. So until the ct `get` operation is taking that much time, it's not going to be noticeable. I would expect that's not going to happen until the size of the ct map is like more than hundreds of thousands or millions.

When we accumulate fees during a block, we validate all the transactions again, then we do basically the same amount of work per transaction (again), to do the add operation. The sorting step is `O(n^2)`. That might get noticeably slow when you have a thousand fee tokens, I'm not really sure. We could implement a better sort that would be `O(n log n)`, it's just complicated. I don't think we need this yet.

Reply:
Yeah %100 in agreement on all points. My comments here are mainly to consider the longer term effects before the long term :). I wish there was some kind of way we could pin certain PRs into a board for later reference so that when (hopefully) we hit the point of being a large, high traffic network with a lot of assets, we have these discussions for reference in future engineering planning.